torch/lib/c10d/test/TCPStoreTest.cpp

```diff
@@ -136,10 +151,222 @@ void testHelper(const std::string& prefix = "") {
   }
 }
 
+void testWatchKeyCallback(const std::string& prefix = "") {
+  // Callback function increments counter of the total number of callbacks that
+  // were run
+  std::promise<int> numCallbacksExecutedPromise;
+  std::atomic<int> numCallbacksExecuted{0};
+  const int numThreads = 16;
+  const int keyChangeOperation = 3;
+  c10d::WatchKeyCallback callback =
+      [&numCallbacksExecuted,
+       &numCallbacksExecutedPromise,
+       &numThreads,
+       &keyChangeOperation](
+          c10::optional<std::string> /* unused */,
+          c10::optional<std::string> /* unused */) {
+        numCallbacksExecuted++;
+        if (numCallbacksExecuted == numThreads * keyChangeOperation * 2) {
+          numCallbacksExecutedPromise.set_value(numCallbacksExecuted);
+        }
+      };
+
+  const int numWorkers = numThreads + 1;
+  auto serverTCPStore = _createServer(numWorkers);
+  auto serverStore =
+      c10::make_intrusive<c10d::PrefixStore>(prefix, serverTCPStore);
+
+  // Each thread will have a client store to send/recv data
+  std::vector<c10::intrusive_ptr<c10d::TCPStore>> clientTCPStores;
+  std::vector<c10::intrusive_ptr<c10d::PrefixStore>> clientStores;
+  for (auto i = 0; i < numThreads; i++) {
+    clientTCPStores.push_back(c10::make_intrusive<c10d::TCPStore>(
+        "127.0.0.1", serverTCPStore->getPort(), numWorkers, false));
+    clientStores.push_back(
+        c10::make_intrusive<c10d::PrefixStore>(prefix, clientTCPStores[i]));
+  }
+
+  // Start watching key on server and client stores
+  std::string internalKey = "internalKey";
+  std::string internalKeyCount = "internalKeyCount";
+  for (auto i = 0; i < numThreads; i++) {
+    serverStore->watchKey(internalKey + std::to_string(i), callback);
+    serverStore->watchKey(internalKeyCount + std::to_string(i), callback);
+    clientStores[i]->watchKey(internalKey + std::to_string(i), callback);
+    clientStores[i]->watchKey(internalKeyCount + std::to_string(i), callback);
```

Comment:
these above two lines are newly added to test registering watch key from client side, right?

Reply:
Yep!